### DISCUSSION TITLE
Error handling

### DISCUSSION THREAD

**Discussion by BigBoulard** [go-resty/resty#579] [Q&A] [answered]
Hi,

I struggle to understand who error handling is working using Resty.
I make a PUT API call that returns a 404 HTTP status code but I can catch it in the code below.

I've tried to use resp.IsError() without success.

I've read some similar topic here which uses the SetError method on a struct (Error or Response depending on the case), but this method is maybe specifically tied to a marshalling error when using the SetResult method.

Do you have king of a simple example to understand the good way to handle error using Resty?
That would be awesome.

Best.

This question has been addressed 2 months ago in the issue <https://github.com/go-resty/resty/issues/560> but maybe it wasn't the good place to post that why I put it here. 

```go
func PutItems( items []string) ([]string, string, rest_errors.RestErr){
  resp, err := client.R().
  SetHeader("Accept", "application/json").
  SetBody(myBody).
  Put(myEndpoint)
  
  if err != nil { // 404 don't happen here apparently
       return nil, rest_errors.NewBadRequestError("API Gateway - http update user cars:" + err.Error())
  }
  
  if resp.IsError() { // 404 is not detected either here
      return nil, errors.New("Unknow error here")
  }

  var createdItems []string
  err2 := json.Unmarshal(resp.Body(), &createdItems)
  if err2 != nil {
     return nil, errors.New("error when unmarshalling items" + err2.Error())
  }
  
  return createdItems, nil
}
```

**Comment by kecci** [comment, marked as answer]
Hi @BigBoulard,

Disclaimer: I would answer this from my project experienced before

There is 3 layer validation for the response from resty client:
1. error response. (it could be error timeout or some else)
2. HTTP Status Code, which is the `resp.StatusCode()` (I would use this for this case)
3. Response Body, which is the `json.Unmarshal(resp.Body(), &yourStruct)`

In your case I would suggest you try to validate the HTTP status code if the statusCode `404`:
```
func PutItems( items []string) ([]string, string, rest_errors.RestErr){
  resp, err := client.R().
  SetHeader("Accept", "application/json").
  SetBody(myBody).
  Put(myEndpoint)
  
  if err != nil { // 404 don't happen here apparently
       return nil, rest_errors.NewBadRequestError("API Gateway - http update user cars:" + err.Error())
  }
  
  if resp.IsError() { // 404 is not detected either here
      return nil, errors.New("Unknow error here")
  }

  // ============= New Validation: Handle this status code =============
  if resp.StatusCode() == 404 { // 404 is should be detected here
      return nil, errors.New("Not Found")
  }
  // ============= New Validation: Handle this status code =============

  var createdItems []string
  err2 := json.Unmarshal(resp.Body(), &createdItems)
  if err2 != nil {
     return nil, errors.New("error when unmarshalling items" + err2.Error())
  }
  
  return createdItems, nil
}
```

I Hope this could be help this discussion.
